File: 백준/Silver/6118. 숨바꼭질/숨바꼭질.py
import sys
from collections import deque

N, M = map(int,sys.stdin.readline().split())
graph = [[] for _ in range(N+1)]
for _ in range(M):
    a, b = map(int,sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

dist = [-1] * (N+1)
dist[1] = 0
queue = deque([1])

while queue:
    node = queue.popleft()
    for nxt in graph[node]:
        if dist[nxt] == -1:
            dist[nxt] = dist[node] + 1
            queue.append(nxt)

max_dist = max(dist)
final_num = [i for i, d in enumerate(dist) if d == max_dist ]
print(f'{final_num[0]} {max_dist} {len(final_num)}')

File: 백준/Silver/6118. 숨바꼭질/숨바꼭질.java
import java.util.*;
import java.io.*;

class Main {
  public static void main(String[] args) throws IOException {
    BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));
    String[] st = br.readLine().split(" ");
    int N = Integer.parseInt(st[0]);
    int M = Integer.parseInt(st[1]);

    List<List<Integer>> graph = new ArrayList<>();

    for (int i =0; i <= N; i++) {
      graph.add(new ArrayList<>());
    }

    for (int i = 0; i < M; i++) {
      String[] num = br.readLine().split(" ");
      int a = Integer.parseInt(num[0]);
      int b = Integer.parseInt(num[1]);
      graph.get(a).add(b);
      graph.get(b).add(a);
    }

    int[] dist = new int[N+1];
    Arrays.fill(dist,-1);

   
    Deque<Integer> queue = new ArrayDeque<>();
    queue.add(1);
    dist[1] = 0;

    while(!queue.isEmpty()) {
      int now = queue.poll();

      for (int next : graph.get(now)) {
        if (dist[next] == -1) {
          dist[next] = dist[now] + 1;
          queue.add(next);
        }
      }
    }

    int count = 0;
    int far = 0;
    int num = 0;
    for (int i =0; i <= N; i++) {
      if (dist[i] > far) {
        far = dist[i];
        num = i;
        count = 1;
      }
      else if (far == dist[i]) {
        count++;
      }
    }

    System.out.println(num + " " + far + " " + count);

  }
} 